### ISSUE TITLE
Batching update in react-hooks

### ISSUE THREAD

**Issue by smmoosavi** [facebook/react#14259]
**Do you want to request a *feature* or report a *bug*?**

bug + improvement

**What is the current behavior?**

Sometimes state updates caused by multiple calls to `useState` setter, are not batched.

sandbox1: https://codesandbox.io/s/8yy0nw2m28
sandbox2: https://codesandbox.io/s/1498n44yr3

Step to reproduce: click on `increment all` text and check the console

diff:
```js
// sandbox1
  const incAll = () => {
    console.log("set all counters");
    incA();
    incB();
  };

// sandbox2
  const incAll = () => {
    setTimeout(() => {
      console.log("set all counters");
      incA();
      incB();
    }, 100);
  };
```
console1:
```
render 
set all counters 
set counter 
set counter 
render 
```
console2:
```
render 
set all counters 
set counter 
render 
set counter 
render 
```

**What is the expected behavior?**

Render function should be called once after multiple calls of  `setState`

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**

react: 16.7.0-alpha.2

**Comment by markerikson** [comment]
This appears to be normal React behavior.  It works the exact same way if you were to call `setState()` in a class component multiple times.

React currently will batch state updates if they're triggered from within a React-based event, like a button click or input change.  It will _not_ batch updates if they're triggered outside of a React event handler, like a `setTimeout()`.

I _think_ there's plans long-term to always batch events, but not sure on the details.

**Comment by gaearon** [comment]
Yeah, this isn’t different from behavior in classes.

You can opt into batching with `ReactDOM.unstable_batchedUpdates(() => { ... })` in the meantime.

**Comment by smmoosavi** [comment]
> Yeah, this isn’t different from behavior in classes.

when we are using `setState` in class, all changed states are batched in one `setState` call.
```js
this.setState({a: 4, b: 5});
```
but when we are using useState hooks, set state functions called seperately:
```js
setA(4);
setB(5);
```
I think it is necessary to have automatic batching for useState hooks.

**Comment by markerikson** [comment]
@smmoosavi : no, React handles both `setState` and `useState` the same way.  If you make multiple calls to either within a React event handler, they will be batched up into a single update re-render pass.

**Comment by smmoosavi** [comment]
>  If you make multiple calls to either within a React event handler, they will be batched up into a single update re-render pass.

if the event handler is an async function, they will not be batched up.

more real-world example:

https://codesandbox.io/s/8lp7y5wj09

```js
  const onClick = async () => {
    setLoading(true);
    setData(null); // batched
    const res = await getData();
    setLoading(false);
    setData(res);
  };
console.log("render", loading, data);
```

output:
```
render false null
// click
render true null
render false null   // <-- bad render call
render false {time: 1545903880314}
// click
render true null
render false null   // <-- bad render call
render false {time: 1545904102818}
```

**Comment by markerikson** [comment]
@smmoosavi : yes, that matches everything I've said already.

React wraps your event handlers in a call to `unstable_batchedUpdates()`, so that your handler runs inside a callback.  Any state updates triggered inside that callback will be batched.  Any state updates triggered _outside_ that callback will _not_ be batched.  Timeouts, promises, and async functions will end up executing outside that callback, and therefore not be batched.

**Comment by rolandjitsu** [comment]
Is there a way though to batch state updates in async calls?

**Comment by markerikson** [comment]
@rolandjitsu : Per Dan's comment earlier:

> You can opt into batching with `ReactDOM.unstable_batchedUpdates(() => { ... })`

**Comment by rolandjitsu** [comment]
@markerikson sorry, I did not pay attention to the comments and I ignored them once I saw `unstable_` 🤦‍♂️

**Comment by takion** [comment]
To solve this problem I use `useReducer` instead of `useState` and my I do my "batching" by grouping my data updates in the reducer ;)

**Comment by gaearon** [comment]
Yep, using the reducer is recommended for anything other than a trivial update.

**Comment by smmoosavi** [comment]
You can replace `useState` with `useReducer` when you are writing hooks from scratch. but when you are using third-party hooks (other libs) which return `setState` you can't use `useReduceer` and you have to use `unstable_batchedUpdates`.

**Comment by Acidic9** [comment]
Spent some time at work today trying to find a good solution to this. Ended up merging my two states (`isAuthenticating` and `user`) into an object to solve this but I would certainly love to see some kind of function such as `useBatch(() => {})` become available to use.

**Comment by nmain** [comment]
I find myself running into a lot of use cases where two state values are _usually_ completely separate, but occasionally might be updated at the same time.  `useReducer` is way too heavy for these and I'd much rather have a stable `batch(() => { changeFoo(...); changeBar(...); })`.

**Comment by andrevinsky** [comment]
Hey, what about doing it with a batch wrapper call, like Redux does it now:
```js
const [ state, dispatch, batch ] = useReducer(reducer, initialState);

batch(() => {
  dispatch(..);
  dispatch(..);
  // etc..
})
```

?

**Comment by smmoosavi** [comment]
@nmain 
I was writing [batched-hook] lib. I think It can help you.

[batched-hook]: https://www.npmjs.com/package/batched-hook

**Comment by nmain** [comment]
Thanks, I'll take a look into that. 👍

**Comment by smmoosavi** [comment]
@gaearon 
how we can detect if already we are inside unstable_batchedUpdates?

**Comment by W3stside** [comment]
@smmoosavi this was explained already, inside `event handlers` it's always wrapped in unsafe batch, else it's not.

**Comment by smmoosavi** [comment]
> @smmoosavi this was explained already, inside event handlers it's always wrapped in unsafe batch, else it's not.

I mean programmatically. something like [this]:

```js
function inBatchedMode() {
  return workPhase !== NotWorking
}
```

[this]: https://github.com/facebook/react/blob/af19e2eb2f8a849a11d2f5fa98d65d46087058e2/packages/react-reconciler/src/ReactFiberScheduler.js#L625

**Comment by rzec-r7** [comment]
@gaearon So you make mention that the only way to make sure that multiple calls to `useState` update handlers will not cause multiple re-renders is by using `ReactDOM.unstable_batchedUpdates()` (which does seem to work), is it possible that this might change / be removed in a patch or minor release?

I assume yes since is it pre-fixed with `unstable_` which make me nervous in using it however it is pretty much the only way I can see being able to use my debounce version of `useState` without causing multiple re-renders (which in the case causes multiple APIs requests which I can't have).

**Comment by markerikson** [comment]
@rzec-r7 : we actually started using this in React-Redux v7, because Dan explicitly encouraged us to.  I believe one of his comments was "Half of Facebook depends on this, and it's the most stable of the 'unstable' APIs".

See https://react-redux.js.org/api/batch

**Comment by Igor10k** [comment]
@gaearon any insight on why batching is still considered `unstable`? Are there any use-cases where it's known to cause problems?

**Comment by hg-pyun** [comment]
This is also likely to happen when the state is changed within a window event.

**Comment by kushalmahajan** [comment]
What about functional version of useState.

```
setState( newItem => [...items, newItem]);
setState(4);
```
So as per @gaearon - 

> Updates will be queued and later executed in the order they were called.

This was some twitter post I remember. Now, what can I assume of this mix of two calls for setState in hooks?

**Comment by pie6k** [comment]
I'm not sure how batching actually works in react.

I understand that if some component uses `useState` twice and I set both inside `batch` call, they'll be batched together and the component will be re-rendered only once.

My question is if there is any benefit of using `batch` if I know I'll call callback that will result in updating many different components.

Eg. let's say I've got some subscription-based hook that other components can use to listen to changes in this channel.

Every time change happens I can call

`subscribtion.listeners.forEach(listener => listener.rerenderWithNewValue(newValue))`

It's very likely, that it will cause many various components to re-render.

In such a case, does wrapping such .forEach call inside batch callback give me any benefit?

**Comment by dominique-mueller** [comment]
I'm having the same question @pie6k has - what if I have a global data stream that many components subscribe to and receive data from? Is there any way to batch those component updates, or would I have to use sth like rAF or setTimeout instead?

**Comment by kalbert312** [comment]
How do I batch updates with a custom reconciler? Using the lib: https://github.com/inlet/react-pixi and `ReactDOM.unstable_batchedUpdates` doesn't work here. I'm guessing just stick with `useReducer`?

**Comment by markerikson** [comment]
Batching is reconciler-dependent.  That's why React-Redux has to semi-depend on importing `unstable_batchedUpdates` from either ReactDOM or React Native, depending on what platform we're running on.  So, if the reconciler you're using doesn't export a form of `unstable_batchedUpdates`, then yeah, you're pretty much out of luck.

**Comment by naisergic** [comment]
> @smmoosavi : yes, that matches everything I've said already.
> 
> React wraps your event handlers in a call to `unstable_batchedUpdates()`, so that your handler runs inside a callback. Any state updates triggered inside that callback will be batched. Any state updates triggered _outside_ that callback will _not_ be batched. Timeouts, promises, and async functions will end up executing outside that callback, and therefore not be batched.

Behaviour is different just check this code sandbox, with useState hook setting of state is happening in a synchronous way and with class component setState is happening is an async way

https://codesandbox.io/s/xenodochial-cherry-4hjte

**Comment by naisergic** [comment]
> @smmoosavi : yes, that matches everything I've said already.
> 
> React wraps your event handlers in a call to `unstable_batchedUpdates()`, so that your handler runs inside a callback. Any state updates triggered inside that callback will be batched. Any state updates triggered _outside_ that callback will _not_ be batched. Timeouts, promises, and async functions will end up executing outside that callback, and therefore not be batched.

Behaviour is different just check this code sandbox, with useState hook setting of state is happening in a synchronous way and with class component setState is happening is an async way

https://codesandbox.io/s/xenodochial-cherry-4hjte

In the above code sandbox I am updating the  state in same manner but one with hook and one with class component but the behaviour is different.

**Comment by ramHruday** [comment]
> I'm having the same question @pie6k has - what if I have a global data stream that many components subscribe to and receive data from? Is there any way to batch those component updates, or would I have to use sth like rAF or setTimeout instead?

In my case I had the same global store to update and delete from various setTimeouts, in that case, setSomevariable wasn't batching..

 the fix which worked for me was using the prev state of
 seSomeVariable((prev) => {prev.add(something); return prev}); 

Actual working is confusing me..but its working ;)

**Comment by bhaidar** [comment]
@markerikson What a useful post! Thank you.

I wonder what happens when a setState() is called inside the setTimeout() handler. The setTimeout is placed inside a new Promise() and the last statement of the handler is a call to resolve() the promise. There is also a then() to run after the promise is fulfilled. 

At what moment of time, will React re-render the component? 

Thanks

**Comment by markerikson** [comment]
@bhaidar : earlier this year I wrote an extensive post on [A (Mostly) Complete Guide to React Rendering Behavior](https://blog.isquaredsoftware.com/2020/05/blogged-answers-a-mostly-complete-guide-to-react-rendering-behavior/), specifically to answer questions like yours :)  I'd encourage you to read through it in detail.

**Comment by bhaidar** [comment]
Thanks @markerikson I will surely read it.

**Comment by dcheung2** [comment]
Am I the only one expect React.useCallback should do state changes in a single batch/transaction for me ?

**Comment by Acidic9** [comment]
@dcheung2 100% agree. There may be some down sides, but coding in React I feel like it's more intuative if useCallback and useEffect did everything in a batch.

**Comment by dcheung2** [comment]
@Acidic9 
To be fair, it won't fix all the problem.
e.g. 
```javascript
React.useCallback(async ()=>{
    setState1(1);
    setState2(1); // likely batched
    await Promise.resolve(2);
    setState1(3); // won't batched
    setState2(3); // won't batched
    await Promise.resolve(4);
    setState1(5);           //non batched
    await Promise.resolve(6);
    setState2(7);          // non batched
 }, []);`
``
but at least it fix half of it. without using anything unstable API.

Or it may helps some eslint rule to detect the problem.

I believe they want the blocking/concurrent mode in new React will eventually fix the cause.

**Comment by dani-mp** [comment]
Does anyone know if the new automatic batching feature in React 18 works also with several `useReducer` dispatches?

Usually, I only dispatch once when something happens in my component, but some people also do something like:
```javascript
dispatch({ type: "ACTION_1" });
dispatch({ type: "ACTION_2" });
```
Would these be batched automatically, resulting in a single render?

**Comment by markerikson** [comment]
@dani-mp : `useReducer` is just another form of React's one-and-only state update implementation, so yes.

**Comment by panta82** [comment]
I just got hit by this footgun.

A codependent state update was working fine while being called from events. But the first time I called it from a resolved promise, it started crashing.

IMO this is a surprising behavior. It should work the same in both situations.

**Comment by markerikson** [comment]
@panta82 the existing behavior is well documented.  But, per https://github.com/reactwg/react-18/discussions/21 , multiple updates in a single event loop tick _outside_ of React's original event handling (ie, in a `.then()` or after an `await`) will now be batched as well.

**Comment by panta82** [comment]
@markerikson This is a very positive change! Exactly what's needed.

As for the documentation, I browsed through the `useState` manual at https://reactjs.org/docs/hooks-state.html , and did a bunch of google searches in the vein of "useState async callback", and found nothing. I am sure there's plenty of documentation about this in GH issues and blog posts, but for me, it got lost in the noise.
